File: leetcode/medium/Add Two Numbers/add-two-numbers.cpp
/**
 * Problem:
 *
 * You are given two non-empty linked lists representing two non-negative integers.
 * The digits are stored in reverse order and each of their nodes contain a single digit.
 * Add the two numbers and return it as a linked list.
 *
 * You may assume the two numbers do not contain any leading zero, except the number 0 itself.
 */

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* head = new ListNode(0);
        ListNode* tail = head;
        ListNode* node1 = l1;
        ListNode* node2 = l2;
        while(node1 != NULL || node2 != NULL)
        {
            if(node1 != NULL)
            {
                tail->val += node1->val;
                node1 = node1->next;
            }
            if(node2 != NULL)
            {
                tail->val += node2->val;
                node2 = node2->next;
            }
            int carry = tail->val / 10;
            tail->val = tail->val % 10;
            ListNode* newTail = new ListNode(carry);
            if(node1 != NULL || node2 != NULL || carry > 0)
            {
                tail->next = newTail;
                tail = newTail;
            }
        }
        return head;
    }
};

File: leetcode/medium/Add Two Numbers/add-two-numbers.py
"""
Problem:

You are given two non-empty linked lists representing two non-negative integers.
The digits are stored in reverse order and each of their nodes contain a single digit.
 Add the two numbers and return it as a linked list.

You may assume the two numbers do not contain any leading zero, except the number 0 itself.
"""

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        list = [0]
        node1 = l1
        node2 = l2
        while(node1 is not None or node2 is not None):
            if(node1 is not None):
                list[-1] += node1.val
                node1 = node1.next
            if(node2 is not None):
                list[-1] += node2.val
                node2 = node2.next
            carry = list[-1] // 10
            list[-1] %= 10
            if(node1 is not None or node2 is not None or carry > 0 ):
                list.append(carry)
        return list
